### DISCUSSION TITLE
웹스토리지, api 를 사용하는 함수의 테스트를 해줘야할까요 ?

### DISCUSSION THREAD

**Discussion by moonheekim0118** [woowacourse-study/2022-code-review-study#43] [Youtube] [answered]
코드를 보니 두가지 경우로 나뉘었어요. (두번째 경우는 많이 없었음)

1. 웹스토리지 / api 를 mocking 해서 웹스토리지 기능을 테스트하는 경우
2. 웹스토리지 / api 를 mocking 하지 않고 웹스토리지 기능을 사용하는 앞,뒤 함수(데이터 가공 및 요청)만 테스트하는 경우

저는 개인적으로 웹스토리지나 api 자체를 mocking 하는 것은 의미가 없다고 생각합니다.
그 Mocking 된 함수도 결국 우리 개발자들이 쓴 것 인데... 테스트를 위한 테스트 같달까요?
그리고 언제든지 웹스토리지나 api가 변경될 수 있는데 그 때 마다 테스트코드의 mock을 대응하려면 너무 불필요한 작업이 늘어날 것 같아요.
물론 테스트 정확도는 높아질 것 같긴해도요.

결국 이러면, 웹스토리지/api 를 직접적으로 다루는 영역과, 거기서 가져온 데이터를 가공하는 영역을 철저히 분리해야겠네요...
뭔가 답정너같지만 여러분들의 의견도 듣고싶어서 디스커션 올려봅니다.
여러분들은 어떻게 생각하시나요?

**Comment by byhhh2** [comment, marked as answer]
공감합니다. 어떤 크루분께서 웹스토리지 Mock객체를 직접 구현하셨더라구요!
그런데 리뷰어님께서 요론 [피드백](https://github.com/woowacourse/javascript-youtube-classroom/pull/79/files/277986fa54f52c2206e32d2e81f541b41c73e60b#r825298920)을 달아주셨습니다.

> 테스트를 위해 Mock 객체를 만드신건 잘하셨습니다
그런데 일반적으로 Mock 객체는 Mockin하는 대상과 동일한 규격이어야 합니다
규격이 다를 경우 테스트에는 성공하지만 실제 코드는 실패하여 테스트 코드의 의미가 없어질 수 있습니다

제가 해당 크루분의 코드를 보았을때, 대부분이 localStorage기능과 비슷하다고 생각을 했습니다. 그런데 동일한 규격이 아니라고 달아주신거 보니, 정말 정말 비슷하게 구현을 해야..테스트의 의미가 있나봅니다. 이 리뷰를 보고 

> 그리고 언제든지 웹스토리지나 api가 변경될 수 있는데 그 때 마다 테스트코드의 mock을 대응하려면 너무 불필요한 작업이 늘어날 것 같아요.

이 부분에 정말 공감했습니다. 

그런데 또 생각해보면, 유튜브 미션의 핵심은 웹스토리지, API의 사용하여 미션을 구현하는 것 같은데, 핵심 부분을 테스트하지 못하면 의미가 있을까라는 생각이 드네요.. 

> 결국 이러면, 웹스토리지/api 를 직접적으로 다루는 영역과, 거기서 가져온 데이터를 가공하는 영역을 철저히 분리해야겠네요...

요 부분이 정답같습니다....

**Comment by wonsss** [comment]
제가 받은 리뷰에서는 로컬스토리지를 Mocking하여 테스트하는 것이 의미가 있다고 말씀해주셨어요.
> <해당 피드백>
[질문]
그런데 이 코드를 다시 보다가 든 생각인데요.
제가 로컬스토리지를 mocking해서 테스트하려는 것은 "저장한 영상을 로컬스토리지에서 불러올 수 있어야 한다."인데, 이처럼 유닛테스트에서 웹스토리지나 API 자체를 Mocking해서 테스트하는 것에 어떤 의미가 있는 것인지 고민이 떠올랐어요.
결국 테스트를 위한 테스트가 될 것 같다는 생각이 들기도 해서요.
해당 테스트가 유닛테스트에서 불필요할 것 같아, 지울까 하는 고민이 드는데, 어떻게 생각하시나요?
[답변]
의미 없지 않습니다!
실제 메소드가 동작하는지 로컬스토리지에 데이터가 있는 상황, 없는 상황을 가정해서 테스트를 할 수 있거든요!
추후엔 오히려 더 테스트 케이스를 늘려서 보강하셔야합니다!
